numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1769,6 +1770,79 @@ array_free(PyObject * v)
     PyObject_Free(v);
 }
 
+static const char *
+_attr_name_as_utf8(PyObject *attr_name)
+{
+    const char *c_attr_name;
+    if (PyUString_Check(attr_name)) {
+        c_attr_name = PyUString_AsUTF8(attr_name);
```

Comment:
Actually, it seems like I was allowing myself to make assumptions in some places, but forbidding myself to in others. `PyUString_AsUTF8` is sufficient for this whole function.

The assumption in question is whether on py2, 
```
#define PyUString_AsUTF8 PyString_AsString
```
is correct. `PyString_AsString(str)` is obviously fine, but `PyString_AsString(unicode)` doesn't seem to actually define an encoding